src/Workspaces/Core/Portable/Workspace/Solution/ProjectDependencyGraph.cs

```diff
@@ -36,9 +36,186 @@ public class ProjectDependencyGraph
+
             _projectIds = projectIds;
             _referencesMap = referencesMap;
+            _lazyReverseReferencesMap = reverseReferencesMap;
+            _transitiveReferencesMap = transitiveReferencesMap;
+            _reverseTransitiveReferencesMap = reverseTransitiveReferencesMap;
+            _lazyTopologicallySortedProjects = topologicallySortedProjects;
+            _lazyDependencySets = dependencySets;
+        }
+
+        internal ProjectDependencyGraph WithAdditionalProjects(IEnumerable<ProjectId> projectIds)
+        {
+            // Adding new projects (that themselves don't have references) cause no changes to any of our direct or transitive references,
+            // and can be trivially added to any topological sort and dependency set information, if we already have it.
+            var newTopologicallySortedProjects = _lazyTopologicallySortedProjects;
+
+            if (!newTopologicallySortedProjects.IsDefault)
+            {
+                newTopologicallySortedProjects = newTopologicallySortedProjects.AddRange(projectIds);
+            }
+
+            var newDependencySets = _lazyDependencySets;
+
+            if (!newDependencySets.IsDefault)
+            {
+                var builder = newDependencySets.ToBuilder();
+
+                foreach (var projectId in projectIds)
+                {
+                    builder.Add(ImmutableArray.Create(projectId));
+                }
+
+                newDependencySets = builder.ToImmutable();
+            }
+
+            return new ProjectDependencyGraph(
+                _projectIds.AddRange(projectIds),
+                referencesMap: _referencesMap,
+                reverseReferencesMap: _lazyReverseReferencesMap,
+                transitiveReferencesMap: _transitiveReferencesMap,
+                reverseTransitiveReferencesMap: _reverseTransitiveReferencesMap,
+                topologicallySortedProjects: newTopologicallySortedProjects,
+                dependencySets: newDependencySets);
+        }
+
+        internal ProjectDependencyGraph WithAdditionalProjectReferences(ProjectId projectId, IEnumerable<ProjectId> referencedProjectIds)
+        {
+            // Ensure we only go over the list once, since we'll be enumerating it many times. This also gives us an easy bail if we can no-op
+            referencedProjectIds = referencedProjectIds.ToList();
+
+            if (!referencedProjectIds.Any())
+            {
+                return this;
+            }
+
+            // Step #1: recompute our references map
+            var newReferencesMap = _referencesMap;
+
+            if (newReferencesMap.TryGetValue(projectId, out var existingReferences))
+            {
+                newReferencesMap = newReferencesMap.SetItem(projectId, existingReferences.Union(referencedProjectIds));
+            }
+            else
+            {
+                newReferencesMap = newReferencesMap.SetItem(projectId, referencedProjectIds.ToImmutableHashSet());
+            }
```

Comment:
this section feels like it could be simplified.  but i'm ok with this as is.

Reply:
It's essentially an immutable multi-map add...we just don't have a helper for that directly. I'm not sure if adding it makes things better...

Reply:
Correct, that is not simpler. :smile: